src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -227,6 +237,74 @@ async def mypy_typecheck_partition(
+                            #
+                            # By default, MyPy uses 2 cache files per source file, which introduces a
+                            # whole slew of race conditions. We can mimize the race conditions by
+                            # using MyPy's SQLite cache. MyPy still has race conditions when using the
+                            # db, as it issues at least 2 single-row queries per source file at different
+                            # points in time (therefore SQLite's own safety guarantees don't apply).
+                            #
+                            # To workaround this we make a copy of the db from the append_only_cache,
+                            # run MyPy on it, then move the updated cache back to the append_only_cache.
+                            # This is multiprocess-safe as mv on the same filesystem is an atomic "rename",
+                            # and any processes copying the "old" file will still have valid file
+                            # descriptors for the "old" file.
+                            #
+                            # There is a chance of multiple processes thrashing on the cache, leaving
+                            # it in a state that doesn't reflect reality at the current point in time,
+                            # and forcing other processes to do potentially done work. This strategy
+                            # still provides a net benefit because the cache is generally _mostly_
+                            # valid (it includes entries for the standard library, and 3rdparty deps,
+                            # among 1stparty sources), and even in the worst case
+                            # (every single file has changed) the overhead of missing the cache each
+                            # query should be small when compared to the work being done of typechecking.
+                            #
+                            # Lastly, we expect that since this is run through Pants which attempts
+                            # to partition MyPy runs by python version (which the DB is independent
+                            # for different versions) and uses a one-process-at-a-time daemon by default,
+                            # multuple MyPy processes operating on a single db cache should be rare.
+
+                            {mkdir.path} -p {run_cache_dir}/{py_version} 2>&1 > /dev/null
+                            {cp.path} {named_cache_dir}/{py_version}/cache.db {run_cache_dir}/{py_version}/cache.db 2>&1 > /dev/null || true
+                            {' '.join(argv)}
+                            EXIT_CODE=$?
+                            {mv.path} {run_cache_dir}/{py_version}/cache.db {named_cache_dir}/{py_version}/cache.db 2>&1 > /dev/null || true
+                            exit $EXIT_CODE
+                        """
+                    ).encode(),
+                    is_executable=True,
```

Comment:
Nit: since this won't have a shebang, it's not executable I don't think.

Reply:
I'll add shebang. Should I try and do bash escaping as well? :thinking:

Reply:
No need to add the shebang: invoking directly with `BashBinary` as you are doing is fine. Sorry I wasn't clear: I was suggesting removing the `is_executable` flag, as it isn't necessary here. No opinion on bash escaping.

Reply:
And would `shell_quote` be sufficient?

Reply:
> And would shell_quote be sufficient?

What for?

Reply:
Ah I was confused and thought I might want the file executable in the sandbox, but now I realize that I don't because `__run.sh` is the runner for it.

@Eric-Arellano I'm wondering if I need to `shell_quote` the formatter strings inside this script.

Reply:
`__run.sh` is not a thing. It's a fake simulation script we only materialize when `--no-process-cleanup`; otherwise, Rust code runs the Process directly.

Reply:
Right, but as a user cracks open the sandbox, I was thinking (incorrectly) they might wanna run this.

Reply:
Without `is_executable` I get `Error launching process: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }`